### PR TITLE
Port redis-py impl and tests to aioredis.

### PR DESCRIPTION
**Pull request by seandstewart**
## What do these changes do?

This is a full port of [redis-py](https://github.com/andymccurdy/redis-py)'s client implementation into an asyncio-friendly implementation.

The approach is essentially a line-by-line re-implementation. The major caveats to this are:

1. Some connection code is simply incompatible and had to be implemented differently.
2. We use black on our code-base, which will create line-count discrepancies between the two sources.
3. I've added a first-pass at type annotations (a lot of work is needed to tighten it up if we want to add mypy to our CI, but it gets us much further along).

Beyond that, all _functional_ logic remains as close to the original, synchronous source as possible.

The motivation behind this is to ease the workload for future contributors and maintainers. By replicating redis-py's implementation, we greatly reduce our own workload, and are able to quickly port new features and bugfixes as they are added to redis-py, including tests.

## Are there changes in behavior for the user?

Yes.

## Related issue number

#822 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is `<Name> <Surname>`.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the [`CHANGES/`](../tree/master/CHANGES) folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example:
   `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by seandstewart** [comment]
@Andrew-Chen-Wang as promised... this is the port from redis-py.

I still need to investigate some intermittent test failures but this is mostly complete.

**Comment by Andrew-Chen-Wang** [comment]
woah this is super cool. I skimmed through the code today and I'll review it later this weekend. I'm a little afraid though that there'll be a lot of breaking changes...? But this is awesome!!!

Additionally, we probably would want to run a benchmark of the port vs. v1.3.1 if there's any significant change (I assume not much since the structure is really pretty much the same). Also note: there are a couple remnants of redis-py in some places like L3827 aioredis/client.py and I'm assuming tested Redis version is v5 for now (i.e. the auth still only allows password)?

**Comment by Andrew-Chen-Wang** [comment]
My biggest concern is this is completely untested in production, and pretty different from the existing aioredis lib. I'm wondering if we should push out a beta version of the current status of the repo, or we should instead just implement this PR and push it out. It'll have a lot of breaking changes, and a lot of people aren't fans of migrations.

**Comment by seandstewart** [comment]
Yeah - I think it’s a valid concern, however we also get the benefit that it’s now much easier for current users of redis-py to migrate their existing code to aioredis, which may help drive adoption and in turn drive community support.

As long as we have a decent alpha/beta/rc period and good migration docs, I think this could be the way to go.

As for breaking changes or the outstanding changes we currently have in master… many of the changes or bug fixes are moot with this PR, and many outstanding features (such as full support for redis 6) are now done. Additionally, master currently contains breaking changes of its own, so I don’t see value in pushing out a beta there if we go this route.

All in all, my main goal in maintaining this library is to make continued maintenance as easy as possible. The loss of the core maintainer of this library exposed a very real problem that happens to codebases with a large amount of custom code and only one author - there is little understanding of the original implementation or the motivations behind it. Such code is always at a high risk of “aging” to a state where no one can touch it or make effective contributions.

I don’t want to see that happen again, and I think the best way to do that is ensure that the code itself can be understood by as many people as possible.

Does that make sense?

**Comment by Andrew-Chen-Wang** [comment]
Was there a reason you needed to add `redis-server --daemonize yes` in d9cbb9b?

**Comment by seandstewart** [comment]
> Was there a reason you needed to add `redis-server --daemonize yes` in [d9cbb9b](https://github.com/aio-libs/aioredis/commit/d9cbb9ba1944a8c1b3f9d45917b981320479dc61)?

Yeah - aioredis has traditionally started all of its own redis-servers internally within the test run. I've been trying to move away from that and instead run the redis as external processes and break out each redis version we test as its own test run.

**Comment by seandstewart** [comment]
So Windows tests are still borked, but I'm wondering if they're of any value considering we're running the tests on a pretty old version of Redis to begin will.

I may mark these as `allowed to fail` and look at putting Windows on Azure pipelines in the near future.

**Comment by Andrew-Chen-Wang** [comment]
Hm I'm still of the opinion most devs aren't even using Redis at this point if they're on Windows; if they were, they would already be using WSL, which would mean they're using the latest version of Redis. At least to me, it feels like unnecessary maintenance and not really worth the effort to test for the lower versions anyways (e.g. redis-py, aredis, and asyncio-redis don't test them).

**Comment by seandstewart** [comment]
Aaaand our tests are green! ✅ 

I'm going to dig into making sure we're exporting our codecov correctly, then update documentation to reflect this major change.

**Comment by popravich** [comment]
OMG, what it has become((((

**Comment by popravich** [comment]
If and when this is merged fill free to rename the library into `redis-py-asyncio`, its just not aioredis any more... (((

**Comment by popravich** [comment]
Now, without a sarcasm.
This is not the way to go!
You literally deleted all aioredis code and cloned redis-py. This library is not a port of something else!
This is not proper place to make redis-py asynchronous, you should have had open this PR in that repository.
You basically **killing work and effort other people put in this project**.

>  Not buffering responses from redis into a byte-stream (as redis-py does). Instead we wrap our futures with deserialization futures, which is hard to debug and leads to ambiguous low-level asyncio errors. Buffering reads into a byte-stream means we can write more discernible and understandable code for our parsing, which now doesn't need to deal with futures directly.

Why not refactoring connection logic to use asyncio protocol not streams?

>  Making the ConnectionPool and Connection objects interchangeable on a single attribute. For simple cases, this makes sense, but for many cases the ambiguity results in potentially bad or unexpected behavior. 

I disagree. First of all, there always was a way to explicitly acquire a connection to use it, there should be documentation and examples how to do it. And AFAIR the case is opposite: this works for most cases.

> The codebase is littered with TODOs acknowledging areas where we need to go back and "pin" a connection to prevent unexpected behavior. A much better pattern here would be to have an explicit connection attribute and always explicitly aquire a connection if necessary.

Partially agree. This how it worked in the very beginning. Agree that connections handling need refactoring but I believe that bringing new ideas can be better than copy-pasting someone else's code. 

> Porting from redis-py also lowers the barrier to implementing new features as Redis advances, as we're now able to source potential solutions from another library.

"implementing" here is just a wrong word, "copying" is the one.

---

All the motivation above looks to me like a reluctance to improve and evolve current aioredis library, this makes me really sad.
Under the hood aioredis is not simple and a person who wants to develop/enhance must deeply understand how it works.

**Comment by seandstewart** [comment]
@popravich -

Welcome back! I'm glad you've returned.

I want to be clear from the start: if you wish to resume maintenance of this library, then I will of course retract this PR and step away. I made it clear in #822 that I was taking this position on in a temporary manner in order to get contributions flowing again.

_This PR is an attempt to make this library maintainable by more than just a single author._ Porting redis-py achieves that goal. It is a pragmatic choice. I don't particularly like all of the decisions in that library, but I do think that using redis-py as an original source means there doesn't have to be a single BDFL maintaining this library, which has been my goal since taking on temporary maintainership. When I made this decision, _the original author hadn't been heard from in over a year_.

Faced with that knowledge, I made the choice to do what you see here today. With this change, continuing contributions to this code-base are much simpler. The redis-py implementation is not only well understood, [it is also officially endorsed by Redis](https://redis.io/clients#python). The aioredis implementation is not well understood by anyone but you. I would say I'm quite familiar with it as well at this point, but that is not an argument for keeping said implementation. That should be a warning sign. When the barrier to maintainership is so high, and the pool of individuals with enough understanding to provide meaningful contributions to its core is so low, that is a signal that the library is doomed to die. And we have proof of that right here.

I don't wish to "kill" others' contributions, but your argument is a sunk cost fallacy. The fact that prior effort and time has been put into a project should not prevent one from changing course when situations change. And they did, drastically.

I do not think the original code was "bad". I actually quite like a lot of it. But that's not why I went this direction and I hope you understand that.

Again, I'll re-iterate, if you wish to resume maintenance and see nothing of value in these changes, then that's fine. I won't fight you simply because I put in the work. You are the original author of this library and I respect that, but I hope you'll understand my reasoning.

**Comment by aviramha** [comment]
This is awesome work @seandstewart!

**Comment by aviramha** [comment]
> > Was there a reason you needed to add `redis-server --daemonize yes` in [d9cbb9b](https://github.com/aio-libs/aioredis/commit/d9cbb9ba1944a8c1b3f9d45917b981320479dc61)?
> 
> Yeah - aioredis has traditionally started all of its own redis-servers internally within the test run. I've been trying to move away from that and instead run the redis as external processes and break out each redis version we test as its own test run.
> 
> It makes debugging test failures much simpler, since it's not one 10K line output per test run

I think you should consider using https://github.com/biocatchltd/yellowbox . It provides a simple wrapper for setting up and tearing containers between tests, and we already have a RedisService class that is easily extensible. This way the tests can run on any system with Docker.

**Comment by Andrew-Chen-Wang** [comment]
I think the biggest problem with this fork is just the huge migration. I feel like taking advantage of the popularity of using aioredis was the point of keeping this as one package rather than converting it into a new package. There are still good portions of aioredis that can inspire some changes in this fork, but I have to agree that the maintainability is quite difficult.

I've used aioredis before for my own package and projects, and I _try_ to help maintain this package, but, to be fair, I also have quite a bit of trouble understanding the code base.

There are a couple options:
1. Make aioredis a little easier to understand to maintain and forget about this fork. @popravich This may be biased, but can **you** point out how you think people could better understand this package besides documentation?
2. Make a new package based on this fork. Erases quite a bit of the popularity, and such adoption is limited and aioredis will probably never release with updates again if maintainer ship does not keep it moving.
3. Use the name of this package in favor of this fork. Adoption is quite clearly high, migration will be a slight pain, but easier maintainability and hopefully more useful PRs to fix any nagging portions.

In my opinion, I actually like option 1 **if** we can identify the pain points of this package (honestly, it's quite a bit like the acquire/release mechanism, magic methods of aenter etc.). But if we can't identify the pain points that basically causing us to not understand how we could fix those TODOs, then the next best option would be option 3.

Thoughts?

**Comment by seandstewart** [comment]
@Andrew-Chen-Wang -

Thanks for the thoughtful response here and sorry for the 5-day delay in responding back. 

I'm obviously biased in my choice, I'd rather rip the band-aid off now (move forward with migrating this codebase) but I do see value in minimizing the changeset and erring on the side of caution. In most any other situation I'd agree to a more conservative approach.

I won't lie though, I really have limited time and won't be around with this frequency forever. My schedule and mental health won't sustain it. My concern with a smaller port is that there is still a barrier to entry for any new maintainer and a piecemeal port means a long-term investment that I'm not sure anyone currently contributing has in them. Maybe that's unfounded, but I'd like to err on the side of caution w.r.t. active maintainership. I know that if all we need to do for new features is watch redis-py for changes and port them over, there's a much much greater likelihood of keeping this library moving. The same can go for the parser logic and the implemented clients.

For connection-related bugs we'll still need to do triage, but thankfully the connections implementation in this PR is easier to grok than the original version so we've still got a win there.

I will defer (a bit) to the will of the community here, with a grain of salt. At the end of the day I want to make we do right by our community and I do think the best way to do that is to keep the codebase maintainable and up-to-date, so whatever we decide to do, let's just be sure we're achieving that goal.

**Comment by seandstewart** [comment]
> > > Was there a reason you needed to add `redis-server --daemonize yes` in [d9cbb9b](https://github.com/aio-libs/aioredis/commit/d9cbb9ba1944a8c1b3f9d45917b981320479dc61)?
> > 
> > 
> > Yeah - aioredis has traditionally started all of its own redis-servers internally within the test run. I've been trying to move away from that and instead run the redis as external processes and break out each redis version we test as its own test run.
> > It makes debugging test failures much simpler, since it's not one 10K line output per test run
> 
> I think you should consider using https://github.com/biocatchltd/yellowbox . It provides a simple wrapper for setting up and tearing containers between tests, and we already have a RedisService class that is easily extensible. This way the tests can run on any system with Docker.

Very nice! I think if/when this PR merges, we can look at what integrating this would mean for our testing.

**Comment by aviramha** [comment]
Any news ?

**Comment by seandstewart** [comment]
> Any news ?

Hey @aviramha - I wanted to wait a few days before moving forward, in the event folks from the community wanted to weigh in.

I'm also working on revamping the documentation and providing a migration guide so that when we land this change there's ample resources to support developers.

**Comment by aviramha** [comment]
> > Any news ?
> 
> 
> 
> Hey @aviramha - I wanted to wait a few days before moving forward, in the event folks from the community wanted to weigh in.
> 
> 
> 
> I'm also working on revamping the documentation and providing a migration guide so that when we land this change there's ample resources to support developers.

Awesome, it's completely understood. I'm a bit paranoid the project will go stale again so was just pinging. If you need any help let me know.

**Comment by popravich** [comment]
I will share my thoughts in a day or two, too much complex work right now.

**Comment by seandstewart** [comment]
> I will share my thoughts in a day or two, too much complex work right now.

Hey @popravich just wanted to check in. I respect that you have work and a life but I'd like to get your input before we move in any direction with this PR. I'd also like to avoid this project becoming stale again.

**Comment by popravich** [comment]
Ok guys, first of all thank you for your contribution to the project.
But I don't think this is the way to go for aioredis, especially when
there already is a redis-py port to asyncio — aredis.

I know that internals of aioredis is hard to understand and I acknowledge
that it must be revised, with current connections management design
(and some other parts) it gets hard to extend the library, especially
with new Redis 6 protocol.

In my opinion we should do the following. First, current issues need to be addressed
to differentiate simple issues/tasks that do not require internals hacking
and are straightforward to fix/implement and would lead library to 1.4 release.
Next, connection management should be redesigned to be ready for new
Redis features which should bring us to aioredis version 2.

I don't want you guys to step back I appreciate your work and I will
help you understand all those internals with TODOs and other stuff
so we can move aioredis to new major release.
I would like to keep nice aioredis features and also make easier to support and develop,
just not with a such a dramatic change.

**Comment by Andrew-Chen-Wang** [comment]
> I don't want you guys to step back I appreciate your work and I will
help you understand all those internals with TODOs and other stuff
so we can move aioredis to new major release.

I'm in favor of this, but I'm also limited on time with school and may need to step back on maintainership. HOWEVER:

> But I don't think this is the way to go for aioredis, especially when
there already is a redis-py port to asyncio — aredis.

aredis is not exactly maintained. I looked around for async redis clients for awhile, and none of them were exactly maintained, so I stuck with aioredis simply because it had the most stars and I thought maintainership would resume under the aio organization. And it did!

Like the synchronous Redis-py lib, we could make aioredis the sole async redis client for the entire Python ecosystem because we 1. have maintainership resumed and 2. have a large community backing. Sorry if that sounds political/inorganic (in growth), but based on everyone's async benchmarks and a different issue where someone said aioredis is 3 times slower than redis-py but the speeds are the same using this port, I think going down this PR's path is also OK.

Yes alternatives are great, but performance and maintainership is key to every library, and none of them have the latter, including aioredis for a long time.

**Comment by popravich** [comment]
> based on everyone's async benchmarks and a different issue where someone said aioredis is 3 times slower than redis-py

It is great that we have such benchmarks so it is possible to identify slow code, but I believe that pipeline/transaction issues can be fixed in terms of aioredis code maybe even reusing redis-py parts/ideas.

**Comment by gnat** [comment]
To give an aioredis users' perspective: It's no secret our community lacks a sustainable, updated, high quality, performant async redis library. I have to give kudos to @seandstewart as he has a sound plan which I for one believe is the quickest (and time may prove, the only sustainable) way to solve this.

I've been closely following and it's concerning our limited resources within the community may be wasted on a sunk cost problem- @popravich has done some amazing work over the years but aioredis' technical debt may be so high that we face a very real danger of project death as a rewrite is likely beyond the scope of what anyone here can sustainably contribute, even collectively. **_This is a worst case scenario for everyone involved._**

The advantages of a redis-py port are very clear and will only get more advantageous as time moves forward, especially as current aioredis falls more behind.

**Comment by popravich** [comment]
Go on with this PR just rename the library and repository to `aioredis-py` so that difference and changes between `aioredis` and releation to `redis-py` is clear and honest. I realized that I would not have enough time to drive the development and it also makes me sad that no one wants to invest time to sort out lib's internals. Its all on you now. I step back. Good luck.

**Comment by seandstewart** [comment]
@popravich - I'm sorry to see you step back but I respect and appreciate your honesty.

@gnat - Thank you for the kind words of support.

I've taken a few days to respond because I wanted to step back and reflect on the long discussion we've had in this PR. I still do believe this is the correct path forward for now, and with the blessing of the community I will continue to push this PR forward. There is still a lot of work to be done on the documentation, which I'm hoping to have done in the coming weeks. At that point, I think we could target a `v2.0.0a1` release with this port.

As @popravich has requested, once this PR is merged, I will also look at renaming the repository to `aioredis-py` to better reflect that this will be a running asyncio-based port of `redis-py`, rather than a custom implementation, as `aioredis` was.

At the end of the day, everyone here wants to see this community and library not only survive, but flourish. I look forward to continued collaboration as we continue our work.

**Comment by Andrew-Chen-Wang** [comment]
Will the package be renamed as well? @seandstewart

**Comment by seandstewart** [comment]
No, I think we’d follow the redis-py convention (yet again). The repo is
`aioredis-py` and the package is `aioredis`.

On March 1, 2021, GitHub Notifications <notifications@github.com> wrote:

> Will the package be renamed as well? @seandstewart
> <https://github.com/seandstewart>
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/aio-

libs/aioredis/pull/891#issuecomment-788153781>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AEVL37UGI2BYBPCFPYQISILTBPJOLANCNFSM4WSLEQIA>.

**Comment by aviramha** [comment]
@seandstewart Any ETA on first release?

**Comment by seandstewart** [comment]
> @seandstewart Any ETA on first release?

I'm hoping to finish up documentation this weekend, then perhaps get an alpha release out by end of next week!

Thanks everyone for your continued patience. I do believe this will enable a much faster iteration cycle moving forward, but it's a Big Change :tm: to get pushed through.

**Comment by acnebs** [comment]
Gonna +1 @popravich here and say it seems like the better place to do this is [aredis](https://github.com/NoneGG/aredis), which is already an asyncio port of `redis-py`. 

To @Andrew-Chen-Wang's point, `aredis` is open source too, so if it's not exactly maintained then you can... help maintain it. But doesn't really make sense to me to have two open source asyncio redis libs that are both just `redis-py` ports. Lots of talk about how "this is the easier way", but seems like the easiest way would've been to just help maintain `aredis`. This on the other hand is duplication of effort already done in the `aredis` lib.

If there is more than one way to solve the problem, having a "sole" asyncio redis lib for python sounds like a bug, not a feature. Trying to morph one lib to look like another lib (for "maintainability") because the first one "has more stars" is silly. Maintain `aredis` and it will soon have more stars.

Also looking at the [benchmarks](https://aredis.readthedocs.io/en/latest/benchmark.html) from the `aredis` repo (which again, is based off `redis-py`), the performance doesn't seem much different between `aioredis` and `aredis`, with `redis-py` being faster than both.

**Comment by Andrew-Chen-Wang** [comment]
@acnebs The problem is who has merging capabilities in these repositories? Only the owners. Why hasn't anyone else taken the torch? The best people can do is make forks and rename the package, but that rarely happens. Well I guess this PR is a special case then.

> If there is more than one way to solve the problem, having a "sole" asyncio redis lib for python sounds like a bug

Diversity's great when there are multiple problems or one with multiple hidden layers software wise. There's really only one for port libraries with one layer of issues. Why do most organizations use psycopg2, redis-py, tensorflow's only Python binding lib, etc.? When they're widely used, they're trusted hence a GitHub star system. It's exactly why there's a market for tech stack websites. People like to use the best there is.

Regarding maintainability, we're all volunteers with school or work. So yes, maintainability is somewhat key. There are easy libraries to understand like Django, redis-py, and aredis where many contributors come to help out. Or there are harder ones like uWebsockets where no one understands it but the owner. And if the owner disappears? Well then the product is just decaying code.

That's just my opinion; I also agree I would like for us to keep the current aioredis code and improve upon it, but to what extent are we all capable of it based on our time and commitment?

**Comment by acnebs** [comment]
Again, if someone (anyone) has the time to maintain a version of `aioredis` that is effectively just an asyncio port of `redis-py`, why not just... have a go with `aredis`? Have you tried asking for perms over there? If not, why not? Seems like that is a much better place to start than here. If the maintainer is unresponsive/AWOL, you can also just *fork* `aredis`. That still seems like a better path forward than this approach.

This is not "maintaining" `aioredis`, this is wholesale replacing it and just trading off the name because it's the most popular lib, as you pointed out yourself.

**Comment by Andrew-Chen-Wang** [comment]
> If not, why not? 

No, because I'm running low on time since I'm an inefficient person, but for others looking: NoneGG's email is in his profile. I've relinquished my role of maintaining in a few repos already but hanging tight here if Sean wants some minor assistance; overlooking another repo brings more responsibilities than I personally think I can handle, and I can't handle it.

**Comment by seandstewart** [comment]
@acnebs -

Thanks for submitting your feedback to this PR. I understand your gut reaction to this change, but I also feel it lacks context.

> it seems like the better place to do this is aredis, which is already an asyncio port of redis-py

You're correct in that aredis _started_ as a direct port of redis-py, but it most definitely not true any longer. The implementation in aredis has diverged significantly since its creation in 2017 and maintenance has dropped off, leaving yet another async redis library without support.

I did not take over maintainership of aioredis with the goal of replacing the implementation. I took over maintainership because there hadn't been an active maintainer for over a year and I wanted to make sure this library and community didn't die out. I spent the end of last year and into the start of this one digging into the code and triaging tickets. During that time I came to the conclusion that it just wasn't possible for the community to realistically support the current implementation of aioredis. If I want to achieve my goal of ensuring this library can be community-supported with minimal barrier to entry, then this is the correct path forward.

You're right, I'm completely replacing the original implementation of aioredis with this change. But software is not sacred. Software is temporary, software is always evolving. We should accept that something like this could occur. A library like aioredis exists for the express purpose of supplying developers with a compliant client for working with Redis in an asyncio context. It does not exist for the maintainer or for the sanctity of its implementation. If we wish to maintain an archive of v1, then we can easily create an archival branch with that implementation, that's what version control is for. There was a period during the course of this PR w the original maintainer looked willing to return and work with us to fix the connection and pipeline implementations, but that period has passed and they have officially stepped back, citing a lack of time to  devote to such work. I understand that decision and support it. This change is an acknowledgement of such limitations.

Of course any of us could fork the library and maintain our own port or set of code. That's not the goal. The goal is to revive this library and ensure the community which _already exists_ continues having access to a high-quality, asyncio-native library for working with Redis.

I hope you understand this decision and can respect the intent, which is to help the community. At the end of the day, the decision is made and I will be moving forward with this change once I feel the documentation is ready to support it.

**Comment by seandstewart** [comment]
I'd also like to add that this isn't the _end_, this is another _beginning_. If the community of contributors grows (which is my hope) then perhaps we change course at that point and bring back some of the patterns from v1 that are desirable. The software will change and evolve with the group that chooses to maintain it, but we need that group to exist first.

**Comment by gjcarneiro** [comment]
Just, *please*, don't publish this aioredis-py as `aioredis` package.  This is essentially a new project, with different API, it should have a different package name.

**Comment by seandstewart** [comment]
Hey @gjcarneiro - 

I understand how you feel. To say this is a large change is an understatement. 

I would like to stress, however, that the client-facing API is largely the same. 

Any differences in the signatures of commands are changes that were either needed to add support for changes in Redis or the result of having a very unique implementation of the command. 

Changes to the connection management to align with the redis-py implementation should not dramatically alter how the client is used or initiated either, as asynchronous contexts are still supported, as will as checking out single connections.

Top-level transactions have not changed, and in fact should have much better performance. 

I don’t initially see the need to publish a new package given the fact that the major version will be bumped, indicating the fact that the API has changed. I also want to allow time before an official release, I won’t rush out a major change like this. There will be pre-releases, testing, and outreach to downstream dependencies.

Does this help assuage your concerns?

**Comment by gjcarneiro** [comment]
It doesn't have the MPSC module (`aioredis.pubsub.Receiver`), which I use in a lot of places.  client.setex() used to accept a float as well as int.  Just two examples of API changes off the top of my head.  I'm sure there are plenty more.

I don't have time or motivation to port old code to a new API.  You're just going to force me to pin aioredis<2 requirement I guess.  Even that is problematic, I have to remember all the modules where I use aioredis, to add this pin.  Others will get more "surprised".

OTOH, kudos for giving it type annotations.  I wanted type hints for a long time in aioredis. :+1: 

It's a cool project, but I wish it was a separate project.

**Comment by seandstewart** [comment]
> It doesn't have the MPSC module (`aioredis.pubsub.Receiver`), which I use in a lot of places. 

I can see now how removing those abstractions could be a pretty large blocker for some. Could you create an issue describing how you use the Channel/Reciever abstraction? I can look at adding it back in, perhaps as a `contrib` or `ext` module to denote that it is an extension over the default implementation.

> client.setex() used to accept a float as well as int. 

Regarding setex, accepting float for that value is actually disingenuous, as the Redis command expects an integer value in seconds: https://redis.io/commands/setex. If you need millisecond precision, psetex should be used, also with a whole  integer value: https://redis.io/commands/psetex. The original implementation [hid this detail](https://github.com/aio-libs/aioredis-py/blob/v1.3.1/aioredis/commands/string.py#L222-L223), but that is an implicit action which could lead to surprising behavior.

**Comment by Andrew-Chen-Wang** [comment]
@seandstewart Do you mind triggering the ReadTheDocs for the alpha release? I'm not sure if publishing the alpha release to PyPi will force everyone to update (e.g. on PyJWT, they published alpha and beta releases in the same format as you did, but dependabot and pip didn't ask users to update) ref: #919

**Comment by seandstewart** [comment]
Yeah, I don't appear to have administrative access to ReadTheDocs and LGTM. I need administrator access to trigger builds and adjust configurations on those services. We just haven't needn't that access till now which is why it went overlooked. @asvetlov should be able to grant access. Additionally, as he's the owner of the project, he can work with @abrookins to get him set up with the correct access. Let's track all this in #822

**Comment by yurymann** [comment]
@seandstewart, @Andrew-Chen-Wang, huge thank you for your efforts! I know I'm a bit late to answer Sean's requests for community verdict, but I have to say that I use aioredis heavily in a mission-critical production project (because it was an asyncio port with the most stars as said above and at the time also seemed relatively maintained) and your way of resurrecting it has my full support. 

I'll feel much more confident about addressing breaking changes with a switch to the maintained port of redis-py than I would about relying on a stale codebase with known bugs and performance issues. As previous commenters, I'm grateful to @popravich and other maintainers of the original codebase for the working version which we currently use. But I was seriously considering moving away from redis to a different message broker/in-memory cache product just because a major part of our code is asyncio-based and with growing user base and system load we would not be able to keep ignoring issues like performance, pipeline transactions and cluster support. I also saw issues with reconnection logic which I didn't get around to understanding better and reporting. 

A note on Windows support: redis 5 and later can't run on Windows, but we still need to use the client lib on Windows as this is where do most coding and debugging.

**Comment by kevr** [comment]
What happened to the [CHANGES/](CHANGES/) additions?

File: tests/ssl_test.py [hunk 1]
```diff
@@ -1,29 +0,0 @@
-import pytest
-
-
-@pytest.mark.asyncio
-async def test_ssl_connection(create_connection, server, ssl_proxy):
```
Comment (webknjaz) [inline tests/ssl_test.py:5]: @seandstewart if you're ever reimplementing the TLS tests, look into integrating the `trustme` library. aiohttp has integration examples in the tests.

**Pull request merged**